### PR TITLE
Add min_hard_negatives argument to the SSDMultiBoxLoss class

### PR DESCRIPTION
**Pull request by djl11**
Add an argument to [SSDMultiBoxLoss](https://github.com/dmlc/gluon-cv/blob/master/gluoncv/loss.py#L98) class constructor, to specify the minimum number of negative examples permitted. This allows the computation of a valid loss in the case of zero positive examples in the batch. Sparsity of positive examples is relatively common in some object detection datasets. The current implementation only allows for a dummy loss in such cases, with no resultant gradients for stepping the weights. This causes .backward() calls to fail unless explicitly handled.

### CODE REVIEW

**Comment by mli** [comment]
Job PR-627-2 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-627/2/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-627-4 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-627/4/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-627-7 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-627/7/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-627-8 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-627/8/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
@djl11 Appreciated for digging into the details, some minor questions regarding this change:

- "This causes training to break unless explicitly handled." What's you observation if loss if completely ignored if no positive sample is present for some images?
- Have you run some experiments to verify this change?

**Comment by djl11** [comment]
`import mxnet as mx
import gluoncv as gcv
import warnings

# set network
net = gcv.model_zoo.get_model('ssd_512_resnet50_v1_custom', pretrained_base=True, classes=('person', 'object'))
net.set_nms()
with warnings.catch_warnings(record=True) as w:
    warnings.simplefilter("always")
    net.initialize()

# define training
net.collect_params().reset_ctx(mx.gpu())
trainer = mx.gluon.Trainer(
    net.collect_params(), 'adam',
    {'learning_rate': 1e-3, 'beta1': 0.9, 'beta2': 0.999, 'epsilon': 1e-3})
mbox_loss = gcv.loss.SSDMultiBoxLoss()

# dummy image data
x = mx.nd.zeros((1,3,512,512),ctx=mx.gpu())

# update network with positive example, without error #
#-----------------------------------------------------#

cls_targets = mx.nd.concat(mx.nd.ones((1,1),ctx=mx.gpu()), mx.nd.zeros((1,6131),ctx=mx.gpu()), dim=1)
box_targets = mx.nd.concat(mx.nd.ones((1,1,4),ctx=mx.gpu()), mx.nd.zeros((1,6131,4),ctx=mx.gpu()), dim=1)
with mx.autograd.record():
    cls_preds, box_preds, _ = net(x)
    first_sum_loss, _, _ = mbox_loss(cls_preds, box_preds, cls_targets, box_targets)
    mx.autograd.backward(first_sum_loss)
trainer.step(1)

print('made first step')

# update network without positive example, with error #
#-----------------------------------------------------#

cls_targets = mx.nd.zeros((1,6132),ctx=mx.gpu())
box_targets = mx.nd.zeros((1,6132,4),ctx=mx.gpu())
with mx.autograd.record():
    cls_preds, box_preds, _ = net(x)
    second_sum_loss, _, _ = mbox_loss(cls_preds, box_preds, cls_targets, box_targets)
    mx.autograd.backward(second_sum_loss)
trainer.step(1)

print('made second step')`

The above is a simple minimal example. With my proposed changes, the second print statement is reached. Without my changes, the second print statement is not reached. Instead, the following error is thrown during the `mx.autograd.backward` call: 

> Cannot differentiate node because it is not in a computational graph. You need to set is_recording to true or use autograd.record() to save computational graphs for backward. If you want to differentiate the same graph twice, you need to pass retain_graph=True to backward.

However, as you pointed out, I could just handle this explicitly in the training loop, by not performing backpropagation in cases where there are no positive examples.

I do still believe that this extra parameter makes the SSDMultiBoxLoss class more general and powerful though, adding flexibility, and preventing users from needing to explicitly handle the special case of no positive examples in the batch. Additionally, if you want identical default behavior as before, we could keep the default value of the `min_hard_negatives = 0` instead of my proposed `min_hard_negatives = 1` if you prefer?

The only thing I haven't checked is whether SSDMultiBoxLoss is used in any tutorials, and if the positioning of the new parameter in the constructor causes any issue for the **kwargs. Although my understanding is these need the key word anyway, and so I can't see how the additional parameter at the end, with a default value, could break anything anywhere else. But do correct me if I'm wrong.

Sorry for quite a long message!
Dan

**Comment by mli** [comment]
Job PR-627-9 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-627/9/index.html 
Code coverage of this PR:  vs. Master:

**Comment by mli** [comment]
Job PR-627-11 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-627/11/index.html 
Code coverage of this PR:  vs. Master:

File: gluoncv/loss.py [hunk 1]
```diff
@@ -133,8 +137,8 @@ def forward(self, cls_pred, box_pred, cls_target, box_target):
             pos_samples = (ct > 0)
             num_pos.append(pos_samples.sum())
         num_pos_all = sum([p.asscalar() for p in num_pos])
-        if num_pos_all < 1:
-            # no positive samples found, return dummy losses
+        if num_pos_all < 1 and self._min_hard_negatives < 1:
+            # no positive samples and no hard negatives, return dummy losses
```
Comment (zhreshold) [inline gluoncv/loss.py]: I think to avoid step error with no gradients, here we can use:
```python
cls_losses = [nd.sum(cp * 0) for cp in cls_pred]
box_losses = [nd.sum(bp * 0) for bp in box_pred]
sum_losses = [nd.sum(cp * 0) + nd.sum(bp * 0) for cp, bp in zip(cls_pred, box_pred)]
return sum_losses, cls_losses, box_losses
```

**Comment by mli** [comment]
Job PR-627-13 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-627/13/index.html 
Code coverage of this PR:  vs. Master:

**Comment by djl11** [comment]
I actually don't think any check is needed at all. Consider the following piece of code:

```
import mxnet as mx
import gluoncv as gcv
import warnings

# set network
net = gcv.model_zoo.get_model('ssd_512_resnet50_v1_custom', pretrained_base=True, classes=('person', 'object'))
net.set_nms()
with warnings.catch_warnings(record=True) as w:
    warnings.simplefilter("always")
    net.initialize()

# define training
net.collect_params().reset_ctx(mx.gpu())
trainer = mx.gluon.Trainer(
    net.collect_params(), 'adam',
    {'learning_rate': 1e-3, 'beta1': 0.9, 'beta2': 0.999, 'epsilon': 1e-3})
mbox_loss = gcv.loss.SSDMultiBoxLoss(min_hard_negatives=0)

# dummy image data
x = mx.nd.zeros((1,3,512,512),ctx=mx.gpu())

# continue stepping network with no examples, and logging loss on single positive example #
#-----------------------------------------------------------------------------------------#

for i in range(100):

    # step loss with no positive or negative examples
    cls_targets = mx.nd.zeros((1,6132),ctx=mx.gpu())
    box_targets = mx.nd.zeros((1,6132,4),ctx=mx.gpu())
    with mx.autograd.record():
        cls_preds, box_preds, _ = net(x)
        no_pos_loss, _, _ = mbox_loss(cls_preds, box_preds, cls_targets, box_targets)
        mx.autograd.backward(no_pos_loss)
    trainer.step(1)

    # log loss for single positive example
    cls_targets = mx.nd.concat(mx.nd.ones((1,1),ctx=mx.gpu()), mx.nd.zeros((1,6131),ctx=mx.gpu()), dim=1)
    box_targets = mx.nd.concat(mx.nd.ones((1,1,4),ctx=mx.gpu()), mx.nd.zeros((1,6131,4),ctx=mx.gpu()), dim=1)
    with mx.autograd.record():
        cls_preds, box_preds, _ = net(x)
        with_pos_loss, _, _ = mbox_loss(cls_preds, box_preds, cls_targets, box_targets)

    print('single pos example loss: ' + str(with_pos_loss[0].asscalar()))
```

In the case where `min_hard_negatives=0`, the trainer steps should never update anything, as there are no positive or negative examples, and nothing to learn. With your proposed changes, the printed loss (for a single positive example) does not change as expected.

However, identical behavior is also observed when this check is removed entirely. Looking through the code, I can see that this is correctly handled by the subsequent for-loop anyway. I would therefore suggest removing this check entirely. But again, possible I missed something. Would be interested to hear your thoughts.

**Comment by mli** [comment]
Job PR-627-15 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-627/15/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
@djl11 Yes, I agree that the output should remain the same. However, speed wise, it's a lot faster to skip the loss entirely.

**Comment by djl11** [comment]
> @djl11 Yes, I agree that the output should remain the same. However, speed wise, it's a lot faster to skip the loss entirely.

True. I have re-added it, and will rebase the commits on my fork. Do you want me to also rebase to master?

File: gluoncv/loss.py (hunk 1 above)
Comment (djl11) [reply gluoncv/loss.py]: I agree, this has now been changed in the commit.

**Comment by zhreshold** [comment]
@djl11 Yes please rebase to master

**Comment by mli** [comment]
Job PR-627-17 is done. 
Docs are uploaded to http://gluon-vision-staging.s3-website-us-west-2.amazonaws.com/PR-627/17/index.html 
Code coverage of this PR:  vs. Master:

**Comment by zhreshold** [comment]
@djl11 Actually this is fine, already merged this, thanks for the contribution!

**Comment by djl11** [comment]
Great, thanks. No problem!
